### ISSUE TITLE
Question about use of AppImage of QPDF

### ISSUE THREAD

**Issue by Nokia808** [qpdf/qpdf#190]
Hi. 
1st I would like to thank you for your very hard efforts !!
2nd I would like to ask some questions about use QPDF as AppImage:

1) I'm on Fedora Linux 26 & already installed QPDF from Fedora repositories (version 6.0.0). Does using of AppImage of QPDF version 8.0.0 can conflict with QPDF installed from repositories or not ?? LibreOffice for example make their AppImage designed in such a way that it will never interfere with installed LibreOffice. Each has it's separated cache folder ..... So, what about AppImage of QPDF ?? Is it safe to be used while QPDF of repositories already installed or not ??

2) I know how to use AppImage of GUI application. But this is 1st time to me to use AppImage for a command line application ! I would like to know what it will opened to me when I run QPDF AppImage ?? After I give AppImage permission to run as a program, then double click it, what will appear to me, GUI terminal or a special command line screen other than my usual GUI terminal ??

**Comment by KurtPfeifle** [comment]
## First Question

An AppImage ***never*** interferes with the ***system*** libraries and files installed by the package manager. The same is of course true for the QPDF AppImage.

However, if the payload application uses a hard-coded configuration file or config directory, it can easily happen that different versions fight over access to these files and may overwrite each other's settings (partially due to the user, who may play around with settings, etc.)

For QPDF this is not a danger at all, because QPDF doesn't use any config file (or if it does, I'd be utterly surprised, because I would never have known it before).

For LibreOffice the developers have ruled out this possibility by defining a config directory path which includes the version string of the current LibreOffice-xxxxxx-x86_64.AppImage.

For most AppImages which are recent enough, you can call them with the *`--appimage-help`* parameter. Try it. If they support it you will see a help screen. There you can learn about how to invoke it with two (different) special options. Use either one:

     ../some/path/to/your.AppImage --appimage-portable-home
     ../some/path/to/your.AppImage --appimage-portable-config

You need to do that only ***once***.

1. Running the first command will create an (initially empty) directory as *`../some/path/to/your.AppImage.home`*, right beside the AppImage itself.

    From now on, the payload application when asking for its *`HOME`* environment variable, it will be told that this is at *`../some/path/to/your.AppImage.home`*. If it wants to write anything to *`$HOME`* config settings and defaults etc., it will use this path. Whenever the AppImage is started, it looks if a directory with the same name as itself, but with the additional *`.home`* extension exists. If it does, it takes it as *`$HOME`*. If not, it uses the "normal" *`$HOME`* and looks for resources there (and may interfere with previous user settings).

1. Running the second command will create an  (initially empty) directory as *`../some/path/to/your.AppImage.config`*, right beside the AppImage itself.

   From now on, the AppImage and its payload application will treat that directory as its XDG_CONFIG_HOME location. (Not all apps use *`XDG_CONFIG_HOME`*, though.)

Of course, you can also manually create these directories. Also, if you delete these directories again, the AppImage will start again using the "standard" *`$HOME`* or *`XDG_CONFIG_HOME`* locations.

Be careful: should you re-name the AppImage, then these directories get disabled. You'll have to rename them exactly as the AppImage's name (but with the added *.home/.config* suffix). The same is true if you move the AppImage to a new directory.

QPDF, however, doesn't care at all about *`$HOME`* nor about *`XDG_CONFIG_HOME`* at all.

## Second Question

What happens if you double click on another command line application on Linux? Does it open a GUI terminal? I don't know, because I've never done this. 95% of my time I'm running SSH sessions into my Linux box(es) and use *`screen`*,  so I'm in a terminal already and start commands from the shell there. And in the 5% of time when I'm sitting in front of a GUI desktop I'm opening a Terminal as the first thing anyway. This is even true when I run Windows or macOS.

So you finally have to find out the answer to your second question yourself.

But nothing "bad" can happen; and if it doesn't start a terminal, you simply open one yourself and run:

     ./qpdf-8.0.0-x64_86.AppImage --help | less
     ./qpdf-8.0.0-x64_86.AppImage [options]

This will behave just as the *`qpdf`* which you may already know. Long-term of course it is easier if you rename the AppImage to something like *`qpdf.ai`* and put it into your *`$HOME/bin`* directory, so that it is in your *`$PATH`*.

If you have a system-installed qpdf available, you can explore the differences:

     qpdf --version
     qpdf.ai --version

The AppImage has documentation and manpages embedded in itself, wheres the system-installed qpdf package has these in other locations on your harddisk.

To get access to the internally embedded goodies like its PDF or HTML manual, run the AppImage like this:

    qpdf.ai --html qpdf-manual.html
    qpdf.ai --pdf qpdf-manual.pdf

This will open these documents in your default HTML or PDF viewer.

To get more info about some other specific options which are uniq to the QPDF AppImage, invoke the specific help screen:

    qpdf.ai --ai-help

**Comment by jberkenbilt** [comment]
Thank you, @KurtPfeifle, for the excellent answer. I will just confirm that qpdf doesn't keep any state or read any configuration files, so it is completely safe to run without any fear of interference with the system package.

**Comment by Nokia808** [comment]
@KurtPfeifle 
@jberkenbilt 

Many thanks !

To make this thread more useful, I will dictate my experience from trying QPDF AppImage here:

1) QPDF AppImage NEVER EVER RUN BY DOUBLE CLICK IT !!! Instead, user need to launch terminal (GUI terminal or tty terminal),

2) No need - at least in my Fedora Linux - to put AppImage of QPDF in $HOME/bin directory for purpose of simplicity. Instead, there is hidden path on my system which is:

home/username/.local/bin

Use Ctrl+H to unmask it, then past QPDF AppImage inside it.

When you put App.Image in above location, terminal will touch it wherever you are located (wherever you open terminal).

It need no root permission to past to it. Just cut/past AppImage of QPDF to this path, then right click AppImage & select "properties", then select "permission" tab, & finally check "Allow executing file as program",

3) to make it easer, change name of AppImage file of QPDF into just qpdf, so that it will become "qpdf.AppImage"

Beware! Do not change .AppImage This is not name of file but it is the extension of file. 

So, no need to change name into qpdf.ai (it will become qpdf.ai.AppImage), but just make it qpdf (it will become qpdf.AppImage)

4) now to use qpdf as AppImage: 

- open your terminal,
- type (as non root):
qpdf.AppImage --[option]

Note: to run your system installed QPDF, just run (as a non root):
qpdf --[option]

To be sure from that run:
qpdf --version
then
qpdf.AppImage --version
& see the difference !
That is all !

**Comment by KurtPfeifle** [comment]
@Nokia808:

If I understand you correctly, you have also the `qpdf` on your system which was installed by the package manager? In `/usr/bin/qpdf` ?

The `PATH` you used to put the AppImage into `/home/username/.local/bin/` is of course also Ok.

With my renaming suggestion I didn't mean `qpdf.ai.AppImage`, but `qpdf.ai`. I assume that your window manager or desktop environment does "hide" the suffix of all or most files from you?

However: whatever works for you is ok. I just wanted to ensure:

* that you can tell which of the two qpdf variants are running, and
* that you do  not have to type (and look at) the long .AppImage suffix all the time...

The AppImage itself does not care at all what name it has. You can even name it *`'q'`*, with no suffix at all, if you want.

**Comment by Nokia808** [comment]
@KurtPfeifle 

Thank you very very much for these data ! Yes, AppImage do not need suffix at all !!!!!!!!!! I delete suffix "AppImage" & make name of AppImage just "qpdf.ai" & it worked O.K !!

I tried to change file extension on Linux & noticed that it became not accessible. However, it was reusable when I re-establish it's suffix. This was for certain file type that I do not remember it exactly ..... On Windows the damage irreversible ..... This make me conservative with AppImage. I did not know before that AppImage not changed by changing or deletion of suffix !!

I will edit my previous post to prevent confusion for peoples read this thread.

Really, really Linux is very great O.S Repeatedly I ask myself: why it is not the 1st OS in common usage? Why most peoples use Windows ??
